pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -18,38 +18,220 @@ package cm
+	"strings"
+	"sync"
 
 	"github.com/golang/glog"
 
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/sets"
 	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha1"
 	"k8s.io/kubernetes/pkg/kubelet/deviceplugin"
 )
 
+// podDevices represents a list of pod to Device mappings.
+type containerToDevice map[string]sets.String
+type podDevices struct {
+	podDeviceMapping map[string]containerToDevice
+}
+
+func newPodDevices() *podDevices {
+	return &podDevices{
+		podDeviceMapping: make(map[string]containerToDevice),
+	}
+}
+func (pdev *podDevices) pods() sets.String {
+	ret := sets.NewString()
+	for k := range pdev.podDeviceMapping {
+		ret.Insert(k)
+	}
+	return ret
+}
+
+func (pdev *podDevices) insert(podUID, contName string, device string) {
+	if _, exists := pdev.podDeviceMapping[podUID]; !exists {
+		pdev.podDeviceMapping[podUID] = make(containerToDevice)
+	}
+	if _, exists := pdev.podDeviceMapping[podUID][contName]; !exists {
+		pdev.podDeviceMapping[podUID][contName] = sets.NewString()
+	}
+	pdev.podDeviceMapping[podUID][contName].Insert(device)
+}
+
+func (pdev *podDevices) getDevices(podUID, contName string) sets.String {
+	containers, exists := pdev.podDeviceMapping[podUID]
+	if !exists {
+		return nil
+	}
+	devices, exists := containers[contName]
+	if !exists {
+		return nil
+	}
+	return devices
+}
+
+func (pdev *podDevices) delete(pods []string) {
+	for _, uid := range pods {
+		delete(pdev.podDeviceMapping, uid)
+	}
+}
+
+func (pdev *podDevices) devices() sets.String {
+	ret := sets.NewString()
+	for _, containerToDevice := range pdev.podDeviceMapping {
+		for _, deviceSet := range containerToDevice {
+			ret = ret.Union(deviceSet)
+		}
+	}
+	return ret
+}
+
+// updateCapacityCallback is called to update ContainerManager capacity when
+// device capacity changes.
+type updateCapacityCallback func(v1.ResourceList)
+
 type DevicePluginHandlerImpl struct {
-	devicePluginManager deviceplugin.Manager
+	sync.Mutex
+	devicePluginManager *deviceplugin.ManagerImpl
+	// allDevices and allocated are keyed by resource_name.
+	allDevices map[string]sets.String
+	allocated  map[string]*podDevices
+	updateCapacity updateCapacityCallback
 }
 
 // NewDevicePluginHandler create a DevicePluginHandler
-func NewDevicePluginHandler() (*DevicePluginHandlerImpl, error) {
+func NewDevicePluginHandler(enabled bool, updateCapacity updateCapacityCallback) (*DevicePluginHandlerImpl, error) {
+	if !enabled {
+		return &DevicePluginHandlerImpl{}, nil
+	}
 	glog.V(2).Infof("Starting Device Plugin Handler")
+	handler := &DevicePluginHandlerImpl{
+		allDevices:          make(map[string]sets.String),
+		allocated:           devicesInUse(),
+		updateCapacity:	     updateCapacity,
+	}
 
-	mgr, err := deviceplugin.NewManagerImpl(pluginapi.DevicePluginPath,
-		func(r string, a, u, d []*pluginapi.Device) {})
+	deviceManagerMonitorCallback := func(resourceName string, added, updated, deleted []*pluginapi.Device) {
```

Comment:
Maybe this would fit more as a method of the `handler`
You would then be able to call NewManagerImpl:
```
deviceplugin.NewManagerImpl(pluginapi.KubeletSocket, handler.deviceManagerMonitorCallback)